src/compiler/checker.ts

```diff
@@ -30122,32 +30185,138 @@ namespace ts {
             return awaitedType || errorType;
         }
 
+        /**
+         * Gets or creates an `awaited T` type for a generic type.
+         *
+         * The "awaited type" of a generic type cannot be determined until it is instantiated. As
+         * a result, an `AwaitedType` for the generic type is created that can be instantiated
+         * or related later.
+         */
+        function getAwaitedTypeForGenericType(type: InstantiableType) {
+            const typeId = "" + type.id;
+            let awaitedType = awaitedTypes.get(typeId);
+            if (!awaitedType) {
+                awaitedType = <AwaitedType>createType(TypeFlags.Awaited);
+                awaitedType.awaitedType = type;
+                awaitedTypes.set(typeId, awaitedType);
+            }
+            return awaitedType;
+        }
+
+        function unwrapAwaitedType(type: Type): Type {
+            return type.flags & TypeFlags.Union ?
+                mapType(type, unwrapAwaitedType) :
+                type.flags & TypeFlags.Awaited ? (<AwaitedType>type).awaitedType : type;
+        }
+
+        /**
+         * Determines whether a type has a callable `then` member.
+         */
+        function isThenableType(type: Type): boolean {
+            const thenFunction = getTypeOfPropertyOfType(type, "then" as __String);
+            return !!thenFunction && getSignaturesOfType(getTypeWithFacts(thenFunction, TypeFacts.NEUndefinedOrNull), SignatureKind.Call).length > 0;
+        }
+
+        /**
+         * Determines whether a type is a generic type whose base constraint could possibly resolve to a different

```

Comment:
Maybe note inline that we know that this is _slightly_ unsound (because `T extends {field: number}` can be instantiated with `Promise<whatever> & {field: number}` or similar), but is unlikely in practice.